### PR TITLE
fallback to SSE

### PR DESCRIPTION
**Pull request by mattkrick**
I'll call this done when SSE works perfectly.
I _think_ it does right now, but I still haven't tested a few things.

Summary
Before, we were using a fork of apollo's client. I finally realized the problem with it was that it was trying to manage both the connection & the graphql protocol. conflating the 2 makes the logic really hard to follow & my first attempt resulted in a really confusing mess.

So, we now have 2 packages:
- trebuchet-client: handles all the connectivity. queues unsent messages, reconnects when disconnected, managers the heartbeat, handles connection errors. sits between graphql & the "trebuchet" (that's what i call a websocket, webrtc peer, event source or poll).
- graphql-trebuchet-client: handles the graphql protocol, creates payloads, etc. sits between relay & trebuchet-client.

BURNDOWN (for SSE & WebSockets):
- Test SSE pings for a long time on the dev server
- change auth (get added to or booted from a team)
- go through welcome wizard
- go through a meeting, including the summary & then back to team dash
- visit an invoice from the org page
- signout & log back in & verify everything was flushed on the client & the server
- remove EventSource from the browser & see it fail gracefully
- kill the eventsource every 3 pings to simulate a firewall crashing the party

TESTS
- [ ] when disconnected, you only get 1 toast saying you were disconnected
- [ ] when reconnected, you only get 1 toast saying you are now reconnected
- [ ] when your computer wakes from sleep, the websocket starts up within 10 seconds (no need for a refresh)  
- [ ] running squid, the fallback to SSE is seamless. the user doesn't even know unless they see the error in the console

### CODE REVIEW

File: src/universal/components/SocketHealthMonitor.js [hunk 1]
```diff
@@ -39,45 +47,18 @@ class SocketHealthMonitor extends Component {
...
-    })
-    dispatch(
-      showError({
-        autoDismiss: 0,
-        title: 'WebSockets Disabled',
-        message: `We weren't able to create a live connection to our server. 
-          Ask your network administrator to enable WebSockets.`
-      })
-    )
+    }, 1000)
```
Comment (mattkrick) [inline src/universal/components/SocketHealthMonitor.js]: if the firewall disconnects things pretty regularly, we don't want to alert the client, so we only send alerts when they've been disconnected for more than 1 second

File: src/universal/utils/constants.js [hunk 2]
```diff
@@ -189,21 +189,6 @@ export const ORG_APPROVAL = 'orgApproval'
 export const TEAM = 'team'
 export const TEAM_MEMBER = 'teamMember'
 
-/*
```
Comment (mattkrick) [inline src/universal/utils/constants.js]: this has all been externalized.

File: webpack/webpack.dev.config.js [hunk 3]
```diff
@@ -45,7 +45,7 @@ module.exports = {
             cacheDirectory: true,
             babelrc: false,
             plugins: [
-              'syntax-object-rest-spread',
+              'transform-object-rest-spread',
```
Comment (mattkrick) [inline webpack/webpack.dev.config.js]: microsoft edge support

File: src/universal/Atmosphere.js [hunk 4]
```diff
@@ -37,127 +82,63 @@ export default class Atmosphere extends Environment {
...
-            // }, 1000)
-          } else {
-            reject(errors && errors[0])
-          }
-        }
-      }
-      this.subscriptionClient.sendMessage(opId, GQL_START, payload)
-    })
+  async promiseToUpgrade () {
+    const trebuchets = [this.trySockets, this.trySSE]
```
Comment (mattkrick) [inline src/universal/Atmosphere.js]: @ackernaut when reviewing, remove `this.trySockets` to test SSE

**Comment by mattkrick** [comment]
@ackernaut ready for review
i don't know for certain how to block websockets on a mac. try installing `squid` from homebrew & then run:
`sudo squid && google-chrome-stable --proxy-server="0.0.0.0:3128"`

**Comment by ackernaut** [comment]
Installing all the things…

###Test No. 1 (in progress)
- [ ] when disconnected, you only get 1 toast saying you were disconnected
- [ ] when reconnected, you only get 1 toast saying you are now reconnected
- [ ] when your computer wakes from sleep, the websocket starts up within 10 seconds (no need for a refresh)  
- [ ] running squid, the fallback to SSE is seamless. the user doesn't even know unless they see the error in the console

**Comment by ackernaut** [comment]
@mattkrick we may need to pair-work testing this one

**Comment by mattkrick** [comment]
what problems ya running into?

if squid isn't working for ya, just remove `this.trySockets` in the comment above & then you should be able to test in 2 tabs

**Comment by ackernaut** [comment]
yeah, squid config is not happy out of the box and I haven’t found a config that I’m happy with yet. For fun I fired up Charles Proxy but I’m not sure what settings will spoof the WS/SSE.

**Comment by mattkrick** [comment]
yeah, let's forget squid then & just try testing by removing `this.trySockets`

**Comment by ackernaut** [comment]
Ok, I’ll try running with only `this.trySSE`

**Comment by mattkrick** [comment]
alrighty, tested this guy on the dev server, it reconnects when the server goes down during a refresh & new & improved, it also reconnects when the client goes down (sleep, wifi disconnect, etc).
